### ISSUE TITLE
Instrument sink batching

### ISSUE THREAD

**Issue by spencergilbert** [vectordotdev/vector#9719]
Following on our buffer instrumentation, we should also look to instrument batches. Given that batches are flushed on a number of conditions additional insight is helpful for operators to optimize their pipelines.

batch sizing and number in-flight, etc

**Comment by tobz** [comment]
Porting over some of the details from a duplicate ticket that I listed, these are the metrics I would want to see come out of any work to add metrics to the batching process:
- (gauge) total number of pending batches
- (gauge) total number of events in pending batches
- (gauge) total size of pending batches
- (histogram) batch TTL (how long a batch lives before being flushed, either due to max limits or timeout)
- (counter) total batches created
- (counter) total batches flushed, by status (did it hit max bytes? max events? timeout?)

(Some of these overlap with @hhromic's comment, obviously.)

**Comment by csjiang** [comment]
hi! any updates here? this type of metric would be super-useful to us.

**Comment by bruceg** [comment]
No, we have not yet prioritized this work.

**Comment by jszwedko** [comment]
Suggestions from a user here: https://github.com/vectordotdev/vector/issues/20284

**Comment by jszwedko** [comment]
> I think also docs for buffering and batching should be extended as it is also very difficult to understand relation between buffer and batch limits.

Agreed, the docs could be expanded. Putting some responses here in the meanwhile.

> If user configures max batch size to be larger than buffer size, what will happen? If there's direct relation, then Vector should throw a warning if buffer size is lower than batch limit and should recommend buffer size being at least same as batch size or in multiples of batch size (eg. batch size * 2 to have some read-ahead from sources).

Sink buffers are decoupled from batching. That is: the buffer just feeds events into the sink as it gets them, and as the sink fetches them. The sink then batches those events in-memory.

There is this diagram that might help: https://vector.dev/docs/reference/configuration/sinks/vector/#buffers-and-batches

> Also how does this work when ARC or concurrency > 1 is being used (then it should be batch size * expected max concurrency).

Again the buffers are decoupled from the in-memory batching so the buffer size doesn't need to be related to the batch size. You can expect one batch to be created per concurrency though.

> And last thing, if user has one source (eg. kafka) and multiple sinks (ES, S3) with different buffer and batch size. Will sink with smaller buffer throttle the other one?

If the small buffer is full, yes, it will apply back-pressure before the larger buffer does. Again, though, batching is done in memory and is decoupled from buffering.

**Comment by fpytloun** [comment]
@jszwedko thank you, that is what I barely remember from Discord discussion some time ago. Still I think it might be beneficial for buffer to be larger than batch size to have some read-ahead depending on the source, correct?
Is there some metric or log message to know when vector applied backpressure due to slow sink. I don't remember seeing any log message for such event 🤔

**Comment by jszwedko** [comment]
> @jszwedko thank you, that is what I barely remember from Discord discussion some time ago. Still I think it might be beneficial for buffer to be larger than batch size to have some read-ahead depending on the source, correct? Is there some metric or log message to know when vector applied backpressure due to slow sink. I don't remember seeing any log message for such event 🤔

The `utilization` metric is the best one that currently exists for identifying back pressure.

Thinking about it a bit more, for in-memory buffers, I think I could see it being beneficial to have the buffer be at least as big as the batch size multiplied by the concurrency so that the next set of requests could be buffered in memory while the current set is in flight.

For disk buffers, I think having it be 2x would be beneficial since data isn't "deleted" from disk buffers until the sink delivers it.

**Comment by timfehr** [comment]
Is there any news of the current status for this feature? This metric would be also really beneficial for us.
